clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v1/provider/agent/KubernetesServerGroupCachingAgent.groovy

```diff
@@ -110,10 +110,20 @@ class KubernetesServerGroupCachingAgent extends KubernetesV1CachingAgent impleme
     }
 
     def jsonResult = objectMapper.writeValueAsString(result.cacheResults)
+    boolean isControllerSetCachingAgentType = false
 
     if (result.cacheResults.values().flatten().isEmpty()) {
       // Avoid writing an empty onDemand cache record (instead delete any that may have previously existed).
       providerCache.evictDeletedItems(Keys.Namespace.ON_DEMAND.ns, [Keys.getServerGroupKey(accountName, namespace, serverGroupName)])
+
+      //Determine if this is the correct agent to delete cache which can avoid double deletion
+      CacheData serverGroup = providerCache.get(Keys.Namespace.SERVER_GROUPS.ns, Keys.getServerGroupKey(accountName, namespace, serverGroupName))
+      if (serverGroup) {
+        String kind = serverGroup.attributes?.get("serverGroup")?.get("kind")
+        if (kind == "StatefulSet" || kind == "DaemonSet") {
+          isControllerSetCachingAgentType = true
```

Comment:
Why not just return here?

Reply:
To handle the case for deleting a replicaset or replication controller.